### PR TITLE
Refactor the login block to allow the whole 2FA and social flows to happen without a page reload

### PR DESCRIPTION
**Pull request by DanReyLop**
For the Woo-branded flow (#41798), we need the whole user login (including Social login and 2FA) to happen in the same page. To accomplish that, the `Login` block now has the following optional props:
- `onTwoFactorRequested`: If present, it will be called with a single parameter (`authType`) when the page must display the 2FA UI. If not present, the current behaviour is preserved (a full-page redirect happens).
- `onSocialConnectStart `: If present, it will be called when the page must display the Social connect UI. If not present, the current behaviour is preserved (a full-page redirect happens).

When showing the 2FA UI, the buttons to switch to another 2FA type ("I want to receive an SMS instead", etc) also triggered a page reload, this refactor also addresses that.

I wanted to keep the changes to the Login block minimal, that's why the code path is a bit convoluted (the Login block calls `onTwoFactorRequested('sms')` only to have its own `twoFactorAuthType` prop change to `sms`, instead of controlling all that as part of the Login block internal state). You can see how this refactor is used in https://github.com/Automattic/wp-calypso/pull/41798/commits/6011f2d59cab55d8e29b0638475e3c90e038e236 .

### CODE REVIEW

**Comment by simison** [comment]
@Automattic/ganon any impact to `/new` flows to be looking/testing for?

**Comment by p-jackson** [comment]
> @Automattic/ganon any impact to `/new` flows to be looking/testing for?

One thing to watch out for is styling, but there's nothing in this PR that would break that. The other is that the `redirect_to` query param should continue to work. I've quickly tested the gutenboarding flow on this branch and it's working fine. To test:

- Log out and go to `/new`
- Quickly go through onboarding flow, site options aren't important for this test
- When you get to the signup page click "Log in" to use an existing account
- You are redirected to the login page
- Log in (should try various kinds of 2fa)
- After successful login should be redirected back to onboarding flow

File: client/blocks/login/two-factor-authentication/two-factor-actions.jsx [hunk 1]
```diff
@@ -59,18 +49,11 @@ class TwoFactorActions extends Component {
...
-				twoFactorAuthType: 'authenticator',
-				isJetpack: this.props.isJetpack,
-				isGutenboarding: this.props.isGutenboarding,
-			} )
-		);
+		this.props.switchTwoFactorAuthType( 'authenticator' );
 	};
 	recordSecurityKey = ( event ) => {
 		event.preventDefault();
-		page( login( { isNative: true, twoFactorAuthType: 'webauthn' } ) );
```
[the remark this answers is not in the document]
Comment (simison) [reply client/blocks/login/two-factor-authentication/two-factor-actions.jsx:73]: @Automattic/ganon worked on Gutenboarding sign-in/up things. :wave:

**Comment by jsnajdr** [comment]
> we need the whole user login (including Social login and 2FA) to happen in the same page

I'd like to better understand this requirement.

The Woo DNA flow will still use URLs starting with `/jetpack/connect`, but the routes will render slightly differently, respecting the `isWooDNA` flag. Is that right?

To log in an existing user, Jetpack Connect redirects to `/log-in` and then back to `/jetpack/connect`. Is that also true for the Woo DNA flow? Or do you plan to render your own login UI inside the flow, reusing the components from `blocks/login`?

When redirecting from `/jetpack/connect` to `/log-in` and back, there will be a full page reload, because `/log-in` lives in its own entrypoint. Are these redirects OK?

When doing `page( login( { ...params } ) )` routing calls to navigate from one `/log-in/*` page to another, these navigations should be handled by the page.js router, i.e., use the history API to navigate and not cause a full page reload. Did you experience a full page reload happening anyways?

There is one place where the login process _must_ do a full page reload: the moment when the login actually happens, i.e., the server sends the session cookies and the state of Calypso changes from "user not logged out" to "user with ID logged in". Is your patch also trying to prevent that?

And, last but not least, why do you need the login to happen in the same page? What breaks when it doesn't?

I'm not saying that your patch is doing anything wrong, just trying to understand the problem.

**Comment by DanReyLop** [comment]
> I'm not saying that your patch is doing anything wrong, just trying to understand the problem.

I'll try to explain. It's fine if you presume I'm doing something wrong, it's a reasonable possibility :)

Let's look first at the existing Jetpack flow (just so we're in the same page). It starts in `/jetpack/connect/authorize`.

If you're logged in, you get the Authorize screen:
Screenshot 2020-06-09 at 10 47 21

If you're logged out, you get the "Sign up" screen:
Screenshot 2020-06-09 at 10 44 44

If, on that "Sign up" screen, you click on the `Already have an account? Sign in` link, a full-page refresh happens and it redirects you to `/log-in/jetpack`:
Screenshot 2020-06-09 at 10 46 28

On that `/log-in/jetpack` screen, if you log in with a 2FA-enabled account, you'll be redirected to `log-in/jetpack/push`, `log-in/jetpack/sms`, etc. As you said, because it's navigation between `/log-in/*` pages, it doesn't cause a full-page refresh.

Now, the new flow is detailed here: paJDYF-Ei-p2
If the user is logged in, nothing changes (except the color palette), it goes to the Authorize page on `/jetpack/connect/authorize`.
If the user is logged out, that same `/jetpack/connect/authorize` page must render a form that can be a login and signup form at the same time. So, the user will enter an email/username. If it corresponds to an existing user, the password field will show. If it doesn't, then the signup form will show.

So, we need the login form to be hosted in a page that's not `/log-in/*`. With these changes, I worry a bit that he browser back-button navigation is lost, but it seems like a reasonable compromise at least for the first iteration.

I'll answer your specific questions now :)

> When doing page( login( { ...params } ) ) routing calls to navigate from one /log-in/* page to another, these navigations should be handled by the page.js router, i.e., use the history API to navigate and not cause a full page reload.

Is there any reason why navigating from any other page to a `log-in/*` page results in a full-page reload? Could we, for example, "whitelist" `/jetpack/connect/*` too so it doesn't full-page reload?

> To log in an existing user, Jetpack Connect redirects to /log-in and then back to /jetpack/connect. Is that also true for the Woo DNA flow?

No.

> Or do you plan to render your own login UI inside the flow, reusing the components from blocks/login?

Yes, that's exactly it :)

> There is one place where the login process must do a full page reload: the moment when the login actually happens, i.e., the server sends the session cookies and the state of Calypso changes from "user not logged out" to "user with ID logged in". Is your patch also trying to prevent that?

That full-page redirect I'm fine with, we aren't trying to prevent that.

**Comment by jsnajdr** [comment]
> So, we need the login form to be hosted in a page that's not `/log-in/*`.

I see. This is the key part of the puzzle that suddenly makes it all click for me 🙂 

The React UI in `blocks/login` is theoretically reusable outside `/log-in/*`, but in practice, it has navigation to `/log-in/*` routes hardcoded, in the form of `page( login( ... ) )` calls. Your patch replaces that with customizable callbacks.

> Is there any reason why navigating from any other page to a `log-in/*` page results in a full-page reload?

That's because `/log-in`, in order to get smaller JS payload and faster loading times, is bundled and shipped as a separate app from the rest of Calypso. In webpack jargon, these apps are called "entrypoints".

> Could we, for example, "whitelist" `/jetpack/connect/*` too so it doesn't full-page reload?

If the navigation between `/jetpack/connect` and `/log-in` is to be handled by a client-side router, without a reload, they need to be a part of the same JS app. I.e., the JS code running on the `/log-in` page needs to "know" how to navigate to `/jetpack/connect` and render its content. But the Jetpack Connect code is not a part of the `/log-in` app. The login page doesn't know what Jetpack Connect is. It can only tell the browser to navigate to `/jetpack/connect`, just like any `<a href>` hyperlink can.

But it doesn't seem we really need to avoid that full-page reload. It's OK to navigate between pages even with reloads. You only need to be careful to store any app state you'll later need to persistent storage and load it again later. Because the JS memory will be lost.

**Comment by DanReyLop** [comment]
> The React UI in blocks/login is theoretically reusable outside /log-in/*, but in practice, it has navigation to /log-in/* routes hardcoded, in the form of page( login( ... ) ) calls.

Yep, I realized too late that the Login block isn't really meant to be reusable outside `/log-in`. I doubled-down instead of looking for alternatives, not sure if that was the correct choice.

> That's because /log-in, in order to get smaller JS payload and faster loading times, is bundled and shipped as a separate app from the rest of Calypso.

Understood 👍 

> But it doesn't seem we really need to avoid that full-page reload. It's OK to navigate between pages even with reloads. You only need to be careful to store any app state you'll later need to persistent storage and load it again later. Because the JS memory will be lost.

I'll try that once the current approach is merged, it has already gone through a lot of code review and testing. A full-page reload, specially when the color palette is not "wpcom blue", is quite jarring (there's a flash of blue/grey background before the Woo styles are loaded). Preserving the browser back navigation seems a worthy goal though.

If we decide to use the `/log-in/*` routes, then the changes on this PR would be redundant, they could be cleaned up, but they don't really affect the existing behaviour.

**Comment by simison** [comment]
> A full-page reload, specially when the color palette is not "wpcom blue", is quite jarring (there's a flash of blue/grey background before the Woo styles are loaded).

That seems like a separate problem? Doesn't that happen also when loading the woo flow first time? Those styles come from the server-side I think and could totally be worked out.

**Comment by DanReyLop** [comment]
> Doesn't that happen also when loading the woo flow first time?

Yes, but having it happen at first load is not as bad as the wrong color flashing in the middle of the flow.

> Those styles come from the server-side I think and could totally be worked out.

That's encouraging to hear :)
We're on a tight schedule here, but it's good to know that adding back full-page navigation to the Woo flow won't result in a poor user experience.

**Comment by vbelolapotkov** [comment]
Maybe it will sound weird but can we render login page under path like `/jetpack/connect/log-in/woo-dna` and allow redirects from it to `/jetpack/connect/log-in/woo-dna/push|sms|etc`?

**Comment by DanReyLop** [comment]
@vbelolapotkov We can't, it's not just an URL matching issue. `/log-in` is on its own Webpack entrypoint, going from anywhere else to `wordpress.com/log-in/*` will always trigger a full-page refresh (if I have understood it correctly).

**Review by vbelolapotkov — APPROVED** [review]
@DanReyLop I've checked the changes and added only minor note on missing prop types. 👍 

Pre-approving for now but following things needed before merging:

1. Rebase on top of master. Changes are messed up at the moment.
2. Resolve this [comment](https://github.com/Automattic/wp-calypso/pull/43059/files#r437209294). Not sure if anything else have to be done there, but would be nice to get final confirmation.
3. Add missing propTypes

File: client/blocks/login/index.jsx [hunk 2]
```diff
@@ -69,6 +69,7 @@ class Login extends Component {
 		twoFactorEnabled: PropTypes.bool,
 		twoFactorNotificationSent: PropTypes.string,
 		isSecurityKeySupported: PropTypes.bool,
+		userEmail: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (DanReyLop) [reply client/blocks/login/index.jsx:72]: Good catch!

**Comment by DanReyLop** [comment]
> If someone extracted it to client/blocks, then it likely was meant to be reusable outside /log-in. It's just you're the first person to actually use it that way 🙂 No wonder you discovered some deficiencies.

Well, lucky me :)

> Making the components customizable via callback props, instead of hard-coding navigations inside them, is a great thing to do 👍

Thanks! Good to know it's not a waste of time even if we decide to go back to full-page navigation.

**Pull request merged**